scanning for elrs network devices with crsf msp frame capability

freasy opened the pull request:
In future ELRS there is CRSF MSP support.

This contains 2 parts because a elrs rx can connect to home wifi or start a access point mode
1. home wifi
in this mode the elrs rx responds to mdns requests and sends 3 responses (A, TXT, SRV), with those i can make a valid tcp port
2. ap mode
in the latest ELRS master - the mdns is fixed, so i rely on that there too

To check if the connection is still valid we have a "down" event from the mdns browser, but thats not realy reliable, so i poll if a http connection can be made every 5 seconds, this check is skipped, when the GUI is connected to a device.

CapnBry commented:
I've mentioned this in Discord but I'll drop it here for the Betaflight folks' info. This should not be using the hostname as the key (which can change). It should use the mDNS TXT records returned from the query `vendor = elrs` and `type = rx` to determine if the target is suitable.

freasy commented:
> I've mentioned this in Discord but I'll drop it here for the Betaflight folks' info. This should not be using the hostname as the key (which can change). It should use the mDNS TXT records returned from the query `vendor = elrs` and `type = rx` to determine if the target is suitable.

Yes, I will implement that once we sorted AP mode out 👍

haslinghuis commented on src/js/port_handler.js:
```diff
@@ -39,18 +42,236 @@ PortHandler.check = function () {
+                    self.elrs_lock = false;
+                }
+            });
+        }, 2000);
```
Seems rather high. This function is already called every 500ms. So it has to fit into this time window.

freasy replied:
I don't want to flood the network. My second idea was to enable AP mode scanning my hand and don't have it on all the time. How do you think about that?

haslinghuis replied:
That might be a better approach.

freasy replied:
ping is removed

haslinghuis commented on src/js/port_handler.js:
```diff
@@ -39,18 +42,236 @@ PortHandler.check = function () {
+        self.elrs_lock = true;
+
+        setTimeout(() => {
+            ping.sys.probe('10.0.0.1', function (isAlive) {
```
Should define a constant for `10.0.0.1`

freasy replied:
ping is removed

blckmn commented:
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> FAIL
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> FAIL
 - `Tested` label found -> FAIL
 - assigned to an approver -> FAIL
 - approver count at least three -> FAIL

CapnBry commented:
While I can't vouch for anything regarding the appropriateness of how this fits into the configurator's scanning pattern, I can say that the ELRS detection from the mDNS responses gets my full approval. This old incorrect comment should be removed though since it no longer does this
```
// check if any of the answers on the response have a name that starts with
// 'elrs_', which indicates that it is an elers device
```

freasy commented:
i refactored to another package which is now event based, so no timer anymore.

With the current ELRS master, the need of pinging a constant is obsolete.

limonspb reviewed (changes requested):
package-lock.json probably should not be added to the PR?
Please also make sure you have only 1 commit per PR (squash, amend etc)
also conflicts

howels commented:
expressLRS v3 RC is approaching rapidly, would be good to get this included in BF configurator so WiFi connection can be tested easily.

haslinghuis commented:
@freasy please rebase and squash

freasy commented:
> package-lock.json probably should not be added to the PR? Please also make sure you have only 1 commit per PR (squash, amend etc)

i removed the lock file

> @freasy please rebase and squash

i rebased current master and squashed to 1 commit

the android part still needs testing - i couldnt do the test, because remote chrome debug wasnt catching the bf app's chromium

sugaarK commented:
@KarateBrot  @daleckystepan  wana has a look at this

haslinghuis commented:
Checked out the branch but it's not working (Hangs at loading as PortHandler can't be initialized) 
Second it produces a yarn.lock file which should be included in the PR.

haslinghuis commented on src/js/port_handler.js:
```diff
@@ -44,6 +142,10 @@ PortHandler.check = function () {
         self.check_serial_devices();
     }
 
+    self.check_elrs_devices();
```
Where is this function defined?

haslinghuis replied:
Making this changes restores PortHandler to normal operation. Did not test ELRS functionality:

```
diff --git a/src/js/port_handler.js b/src/js/port_handler.js
index a9230135..dab965bb 100644
--- a/src/js/port_handler.js
+++ b/src/js/port_handler.js
@@ -1,40 +1,5 @@
 'use strict';
 
-const http = require("http");
-const bonjour = require('bonjour')();
-
-const MDNS_INTERVAL = 10000;
-const TCP_CHECK_INTERVAL = 5000;
-const TCP_TIMEOUT = 2000;
-
-let mdnsBrowser = null;
-if (GUI.isCordova()) {
-    mdnsBrowser = {
-        services: [],
-    };
-
-    const zeroconf = cordova.plugins.zeroconf;
-    zeroconf.registerAddressFamily = 'ipv4'; // or 'ipv6' ('any' by default)
-    zeroconf.watchAddressFamily = 'ipv4'; // or 'ipv6' ('any' by default)
-    zeroconf.watch("_http._tcp.", "local.").subscribe(result => {
-        const action = result.action;
-        const service = result.service;
-        if (['added', 'resolved'].includes(action)) {
-            console.log("Zeroconf Service Changed", service);
-            mdnsBrowser.services.push({
-                addresses: service.ipv4Addresses,
-                txt: service.txtRecord,
-                fqdn: service.hostname,
-            });
-        } else {
-            console.log("Zeroconf Service Removed", service);
-            mdnsBrowser.services = mdnsBrowser.services.filter(s => s.fqdn !== service.hostname);
-        }
-    });
-} else {
-    bonjour.find({ type: 'http' });
-}
-
 const TIMEOUT_CHECK = 500 ; // With 250 it seems that it produces a memory leak and slowdown in some versions, reason unknown
 
 const usbDevices = { filters: [
@@ -50,6 +15,7 @@ const PortHandler = new function () {
     this.port_available = false;
     this.showAllSerialDevices = false;
     this.showVirtualMode = false;
+    this.mdnsBrowser = null;
 };
 
 PortHandler.initialize = function () {
@@ -66,7 +32,41 @@ PortHandler.initialize = function () {
 
     const self = this;
 
-    mdnsBrowser?.on('down', function (service) {
+    const http = require("http");
+    const bonjour = require('bonjour')();
+    
+    const MDNS_INTERVAL = 10000;
+    const TCP_CHECK_INTERVAL = 5000;
+    const TCP_TIMEOUT = 2000;
+    
+    if (GUI.isCordova()) {
+        self.mdnsBrowser = {
+            services: [],
+        };
+    
+        const zeroconf = cordova.plugins.zeroconf;
+        zeroconf.registerAddressFamily = 'ipv4'; // or 'ipv6' ('any' by default)
+        zeroconf.watchAddressFamily = 'ipv4'; // or 'ipv6' ('any' by default)
+        zeroconf.watch("_http._tcp.", "local.").subscribe(result => {
+            const action = result.action;
+            const service = result.service;
+            if (['added', 'resolved'].includes(action)) {
+                console.log("Zeroconf Service Changed", service);
+                self.mdnsBrowser.services.push({
+                    addresses: service.ipv4Addresses,
+                    txt: service.txtRecord,
+                    fqdn: service.hostname,
+                });
+            } else {
+                console.log("Zeroconf Service Removed", service);
+                self.mdnsBrowser.services = self.mdnsBrowser.services.filter(s => s.fqdn !== service.hostname);
+            }
+        });
+    } else {
+        bonjour.find({ type: 'http' });
+    }
+
+    self.mdnsBrowser?.on('down', function (service) {
         console.log('Lost an HTTP server:', service);
         let currentPorts = [];
         if (self.initialPorts && self.initialPorts.length > 0) {
@@ -86,10 +86,10 @@ PortHandler.initialize = function () {
                 const tcpPorts = self.initialPorts.filter(p => p.path.startsWith('tcp://'));
                 tcpPorts.forEach(function (port) {
                     const removePort = () => {
-                        mdnsBrowser?._removeService(port.fqdn);
+                        self.mdnsBrowser?._removeService(port.fqdn);
 
                         if (GUI.isCordova()) {
-                            mdnsBrowser.services = mdnsBrowser.services.filter(s => s.fqdn !== port.fqdn);
+                            self.mdnsBrowser.services = self.mdnsBrowser.services.filter(s => s.fqdn !== port.fqdn);
                         }
                     };
                     http.get({
@@ -123,7 +123,7 @@ PortHandler.initialize = function () {
 
     self.mdns_timer = setInterval(() => {
         if (!GUI.connected_to) {
-            mdnsBrowser?.update();
+            self.mdnsBrowser?.update();
         }
     }, MDNS_INTERVAL);
 
@@ -158,7 +158,7 @@ PortHandler.check_serial_devices = function () {
 
         let currentPorts = [
             ...cp,
-            ...(mdnsBrowser?.services?.filter(s => s.txt.vendor === 'elrs' && s.txt.type === 'rx')
+            ...(self.mdnsBrowser?.services?.filter(s => s.txt.vendor === 'elrs' && s.txt.type === 'rx')
                 .map(s => s.addresses.map(a => ({
                     path: `tcp://${a}`,
                     displayName: `${s.txt.target} - ${s.txt.version}`,
```

haslinghuis commented on package.json:
```diff
@@ -66,6 +68,8 @@
     "jquery-ui-npm": "^1.12.0",
     "lru_map": "^0.3.3",
     "marked": "^4.0.17",
+    "marked": "^4.0.10",
```
please omit as it's updated to 4.0.17

freasy replied:
yeah i missed that, i already deleted that line.

chmelevskij commented on package.json:
```diff
@@ -75,7 +79,7 @@
     "three": "~0.97.0",
     "universal-ga": "^1.2.0",
-    "vue": "2.6.12"
+    "vue": "^2.6.14"
```
please keep the pinned dependencies as they are not related to this PR

freasy replied:
if you install it from master, 2.6.12 vue gets installed together with 2.6.14 vue-template-compiler. This just fixed both to the same version, so nothing breaks.

chmelevskij commented on package.json:
```diff
@@ -139,7 +143,7 @@
     "vinyl-source-stream": "^2.0.0",
     "vue-loader": "^15.9.8",
-    "vue-template-compiler": "^2.6.12",
+    "vue-template-compiler": "^2.6.14",
```
same here, don't see the reason to bump with this pr

freasy replied:
see above

chmelevskij commented on src/js/port_handler.js:
```diff
@@ -1,5 +1,50 @@
+};
+
+function isMobile() {
+    return ["Android", "Mac"].includes(GUI_checkOperatingSystem());
```
why Mac is in `isMobile`?

freasy replied:
this is indeed wrong! I was didnt looked close and thought this is iOS, i deleted it.

chmelevskij commented on src/js/port_handler.js:
```diff
@@ -52,7 +143,19 @@ PortHandler.check = function () {
+        let currentPorts = [
+            ...cp,
+            ...(mdnsBrowser?.services?.filter(s => s.txt.vendor === 'elrs' && s.txt.type === 'rx')
+                .map(s => s.addresses.map(a => ({
+                    path: `tcp://${a}`,
+                    displayName: `${s.txt.target} - ${s.txt.version}`,
+                    fqdn: s.fqdn,
+                    vendorId: 0,
+                    productId: 0,
+                }))).flat() ?? []),
```
another approach
```suggestion
                .flatMap(s => s.addresses.map(a => ({
                    path: `tcp://${a}`,
                    displayName: `${s.txt.target} - ${s.txt.version}`,
                    fqdn: s.fqdn,
                    vendorId: 0,
                    productId: 0,
                }))) ?? []),
```

chmelevskij replied:
up to yourself 😉 just an alternative

haslinghuis commented on src/js/port_handler.js:
```diff
@@ -1,5 +1,50 @@
+    services: [],
+};
+
+function isMobile() {
+    return "Android" === GUI_checkOperatingSystem();
+}
+
+if (isMobile()) {
```
```suggestion
if (GUI.isCordova()) {
```

freasy replied:
This does not work, I had it this way earlier, but when port_handler initializes, the GUI object is not there. This is why I extracted the other function.

haslinghuis commented:
@freasy by moving some initialization from global scope to modules initialization function with this diff you can use `GUI.isCordova()` 
[diff.txt](https://github.com/betaflight/betaflight-configurator/files/9041985/diff.txt)

haslinghuis commented on src/js/port_handler.js:
```diff
@@ -168,9 +271,7 @@ PortHandler.detectPort = function(currentPorts) {
-            if (result.last_used_port.includes('tcp')) {
-                self.portPickerElement.val('manual');
-            } else if (newPorts.length === 1) {
+            if (newPorts.length === 1) {
```
`result.last_used_port` is not used anymore. Please remove.

haslinghuis commented on src/js/serial_backend.js:
```diff
@@ -91,7 +91,7 @@ function initializeSerialBackend() {
                     }
 
-                    mspHelper.setArmingEnabled(true, false, onFinishCallback);
+                    mspHelper?.setArmingEnabled(true, false, onFinishCallback);
```
Why is `setArmingEnabled `not available here?

freasy replied:
i have no idea, but if i dont do that, the app wont start port_handler

haslinghuis replied:
@chmelevskij any idea?

chmelevskij replied:
https://github.com/betaflight/betaflight-configurator/blob/54738ef833d2a5819caca20ee3a5fc613cc7c617/src/js/serial_backend.js#L3

It's not defined at first, so might be something is async and race condition breaks it.

haslinghuis replied:
```suggestion
                    mspHelper.setArmingEnabled(true, false, onFinishCallback);
```
This no longer is needed.

haslinghuis commented:
@freasy would like to continue working on this (need a rebase).

freasy commented:
> @freasy would like to continue working on this (need a rebase).

Yes of course! I will rebase it, what is missing so far, I think Android gut tested and verified working right?

howels commented:
> > @freasy would like to continue working on this (need a rebase).
> 
> Yes of course! I will rebase it, what is missing so far, I think Android got tested and verified working right?

Last time I tested Android this was not working. I'll try the latest code.

Asizon commented:
> > > @freasy would like to continue working on this (need a rebase).
> > 
> > 
> > Yes of course! I will rebase it, what is missing so far, I think Android got tested and verified working right?
> 
> Last time I tested Android this was not working. I'll try the latest code.

For me is working really, fine, but first we need to solve some not related issues in android.

haslinghuis commented:
@freasy please rebase so people are able to test.

freasy commented:
> @freasy please rebase so people are able to test.

@haslinghuis i rebased the current this branch onto the current master, i cant test it right now, i can do that later.

haslinghuis commented on src/js/helper.js:
```diff
@@ -0,0 +1,15 @@
+function GUI_checkOperatingSystem() {
+    if (navigator.appVersion.indexOf("Win") !== -1) {
+        return "Windows";
+    } else if (navigator.appVersion.indexOf("Mac") !== -1) {
+        return "MacOS";
+    } else if (navigator.appVersion.indexOf("Android") !== -1) {
+        return "Android";
+    } else if (navigator.appVersion.indexOf("Linux") !== -1) {
+        return "Linux";
+    } else if (navigator.appVersion.indexOf("X11") !== -1) {
+        return "UNIX";
+    } else {
+        return "Unknown";
+    }
+}
```
`navigator.appVersion` has been deprecated. We already have this function in gui.js but it is not an esm module yet.

```
function GUI_checkOperatingSystem() {
    return navigator?.userAgentData?.platform || 'Android';
}
```

freasy replied:
removed helper.js, the reference conflict got removed and its not needed in port_handler anymore.

chmelevskij commented on src/js/helper.js:
```diff
@@ -0,0 +1,15 @@
+function GUI_checkOperatingSystem() {
+    if (navigator.appVersion.indexOf("Win") !== -1) {
```
https://github.com/betaflight/betaflight-configurator/blob/c8d0490b2ffa66615ea37da0e9a7282af8d01077/src/js/gui.js#L77

freasy replied:
removed helper.js, the reference conflict got removed and its not needed in port_handler anymore.

chmelevskij commented on src/main.html:
```diff
@@ -84,6 +84,7 @@
     <script type="text/javascript" src="./js/utils/VtxDeviceStatus/TrampDeviceStatus.js"></script>
     <script type="text/javascript" src="./js/utils/VtxDeviceStatus/SmartAudioDeviceStatus.js"></script>
     <script type="text/javascript" src="./js/utils/VtxDeviceStatus/Rtc6705DeviceStatus.js"></script>
+    <script type="text/javascript" src="./js/helper.js"></script>
```
Wouldn't need this if https://github.com/betaflight/betaflight-configurator/blob/c8d0490b2ffa66615ea37da0e9a7282af8d01077/src/js/gui.js#L77 is used.

freasy commented:
> @freasy please can you review this now eLRS v3 is released?

i am currently on it

freasy commented:
@howels @haslinghuis @chmelevskij

i removed helper.js, it seems like its not needed anymore. And i changed some bonjour code, because the "update" call was not functioning anymore (dont know why) but now its fine again.

chmelevskij commented on src/js/port_handler.js:
```diff
@@ -29,6 +32,94 @@ PortHandler.initialize = function () {
+                const action = result.action;
+                const service = result.service;
+                if (['added', 'resolved'].includes(action)) {
+                    console.log("Zeroconf Service Changed", service);
```
Would any of these logs be useful for the UI layer? E.G. https://github.com/betaflight/betaflight-configurator/blob/master/src/js/msp/MSPHelper.js#L2940

freasy replied:
totaly, i will change it, so that those logs go to the GUI as well.

haslinghuis replied:
Please change in follow up patch.

chmelevskij reviewed (approved):
Just the question about the GUI.logs.

howels commented:
Tested on Android and it works. For some reason it wouldn't show the FC on the first launch but after Force-Stopping and restarting the app it's working so perhaps it just takes a second to do the mDNS lookup.

freasy commented:
@howels 
> Tested on Android and it works. For some reason it wouldn't show the FC on the first launch but after Force-Stopping and restarting the app it's working so perhaps it just takes a second to do the mDNS lookup.

yes mDNS is a b**** specially on android, so if it works sometimes, thats already a big win! I am glad, that it works, since i could not test the code myself.

CapnBry commented:
Don't know if you're looking for more "Me too"s, but I just installed the artifact on my Samsung Galaxy S10e and it discovered my ELRS RX no problem on my home wifi network, then switched over to put the RX into AP mode and it discovered properly on that network as well.

My crap eyes thank you for not making me mistype tcp://10.0.01 by mistake 100x before realizing the gray on gray text is missing a pixel!

ASDosjani commented:
@freasy I tested it, it's working on my phone but not on my pc. I'm using win11.
